### PR TITLE
[Merged by Bors] - feat(Combinatorics/SimpleGraph): add `Fintype` instance for `Path`

### PR DESCRIPTION
**Pull request by niklasmohrin**
---

I tried to golf the implementation a bit, I would have expected the following to work:

```lean
complete p := by
  simp
  exact ⟨p.val.length, p.prop.length_lt, p.val, p.prop, rfl⟩
```

... but for some reason only the longer version compiles. Any ideas?

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Combinatorics/SimpleGraph/Connectivity.lean
Comment (eric-wieser) [inline Mathlib/Combinatorics/SimpleGraph/Connectivity.lean:2619]: This is going to be inefficient for computation, I think; something like
```suggestion
  elems := (range (Fintype.card V)).disjiUnion
    (fun n ↦ (univ (α := {p : G.Walk u v | p.IsPath ∧ p.length = n})).map
      ⟨fun p ↦ { val := p.val, property := p.prop.left }, sorry⟩) sorry
```
would behave much better.

Of course, this has the downside that you have two proofs to fill!

File: Mathlib/Combinatorics/SimpleGraph/Connectivity.lean
Comment (eric-wieser) [inline Mathlib/Combinatorics/SimpleGraph/Connectivity.lean:2626]:
```suggestion
    simp only [mem_biUnion, mem_range, mem_image, mem_univ, Subtype.exists, true_and]
    exact ⟨p.val.length, p.prop.length_lt, p.val, ⟨⟨p.prop, rfl⟩, rfl⟩⟩
```

File: Mathlib/Combinatorics/SimpleGraph/Connectivity.lean
Comment (niklasmohrin) [reply Mathlib/Combinatorics/SimpleGraph/Connectivity.lean:2619]: Sounds good. Out of curiosity, where would this inefficiency show? Is it just that we are avoiding the two extra deduplication steps during construction of the `Finset`?

File: Mathlib/Combinatorics/SimpleGraph/Connectivity.lean
Comment (niklasmohrin) [inline Mathlib/Combinatorics/SimpleGraph/Connectivity.lean:2665]: Here it complained with just `mem_univ _`: `failed to synthesize Fintype ↑fun p ↦ p.IsPath ∧ p.length < Fintype.card V` - I guess it is fine like this?

File: Mathlib/Combinatorics/SimpleGraph/Connectivity.lean
Comment (eric-wieser) [reply Mathlib/Combinatorics/SimpleGraph/Connectivity.lean]: It would come when evaluating some computable sum over all paths, which would then perform all this deduplication at runtime. It would also appear if you evaluate something like `example : Fintype.card (G.Path _ _) = 10 = rfl` for some computable `G`.

File: Mathlib/Combinatorics/SimpleGraph/Connectivity.lean
Comment (IvanRenison) [inline Mathlib/Combinatorics/SimpleGraph/Connectivity.lean:2562]:
```suggestion
      have hl : p.length = l := p.mem_finsetWalkLength_iff_length_eq.mp (hsl hp)
```

File: Mathlib/Combinatorics/SimpleGraph/Connectivity.lean
Comment (IvanRenison) [inline Mathlib/Combinatorics/SimpleGraph/Connectivity.lean:2563]:
```suggestion
      have hl' : p.length = l' := p.mem_finsetWalkLength_iff_length_eq.mp (hsl' hp)
```

File: Mathlib/Combinatorics/SimpleGraph/Connectivity.lean
Comment (IvanRenison) [inline Mathlib/Combinatorics/SimpleGraph/Connectivity.lean:2604]:
```suggestion
    rw [← Finset.mem_coe, coe_finsetWalkLengthLt_eq]
```

File: Mathlib/Combinatorics/SimpleGraph/Connectivity.lean
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/Connectivity.lean:2587]: @henkkuli, thoughts?

File: Mathlib/Combinatorics/SimpleGraph/Connectivity.lean
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/Connectivity.lean:2581]:
```suggestion
def finsetWalkLengthLT (n : ℕ) (u v : V) : Finset (G.Walk u v) :=
```

File: Mathlib/Combinatorics/SimpleGraph/Connectivity.lean
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/Connectivity.lean:2597]: Iffs take implicit arguments
```suggestion
theorem mem_finsetWalkLengthLT {n : ℕ} {u v : V} {p : G.Walk u v} :
```

File: Mathlib/Combinatorics/SimpleGraph/Connectivity.lean
Comment (YaelDillies) [reply Mathlib/Combinatorics/SimpleGraph/Connectivity.lean:2688]: I think it's fixed if you do...

File: Mathlib/Combinatorics/SimpleGraph/Connectivity.lean
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/Connectivity.lean:2683]: ...this
```suggestion
  elems := (univ (α := { p : G.Walk u v | p.IsPath ∧ p.length < Fintype.card V })).map
```

**Comment by YaelDillies** [comment]
Can you please merge master?

File: Mathlib/Combinatorics/SimpleGraph/Connectivity.lean
Comment (niklasmohrin) [reply Mathlib/Combinatorics/SimpleGraph/Connectivity.lean:2597]: Ah, I just adapted the lemma from https://github.com/leanprover-community/mathlib4/blob/bc3ab3eddef1b92b87d321f25c74899502c25239/Mathlib/Combinatorics/SimpleGraph/Connectivity/WalkCounting.lean#L112 - should we change it there too then?

File: Mathlib/Combinatorics/SimpleGraph/Connectivity.lean
Comment (niklasmohrin) [reply Mathlib/Combinatorics/SimpleGraph/Connectivity.lean:2683]: Yup, that worked! :) 

Any idea why one works better than the other?

Comment (YaelDillies) [reply Mathlib/Combinatorics/SimpleGraph/Connectivity.lean]: `{p : G.Walk u v | p.IsPath ∧ p.length < Fintype.card V}` is a set, `{p : G.Walk u v // p.IsPath ∧ p.length < Fintype.card V}` is a subtype. You were using the latter as a set, which doesn't really work (well, it does, but only if you spend some effort convincing Lean so)

**Comment by niklasmohrin** [comment]
Ah, one advantage of keeping `Network.loopless` would be that once `SimpleGraph V extends DiGraph V`, we could have `UndirectedNetwork V R extends Network V R, SimpleGraph V` - maybe that would be too restricting though, as there could theoretically also be non-simple graphs employed with a symmetric capacity function.

I suppose the most flexible design to have all kinds of networks carry along all kinds of graphs would be `structure Network (G : Digraph V) (R : Type u_R)`. However, this would defeat the main purpose for the move to the `Network extends Digraph` design :thinking:

What should we do?

File: Mathlib/Combinatorics/SimpleGraph/Connectivity/WalkCounting.lean
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/Connectivity/WalkCounting.lean:118]:
```suggestion
/-- The `Finset` of walks from `u` to `v` with length less than `n`. See `SimpleGraph.finsetWalkLength` for ???. -/
```
where `???` is something you should fill in

**Review by YaelDillies — APPROVED** [review]
maintainer merge

**Comment by niklasmohrin** [comment]
@YaelDillies what should we do about https://github.com/leanprover-community/mathlib4/pull/13810#discussion_r1723735522 ?

**Comment by YaelDillies** [comment]
Oh, Henrik must be busy for now. We will update his project once this PR has merged.

**Comment by niklasmohrin** [comment]
And what about mem_finsetWalkLengthLt_iff_length_lt and the implicit arguments?

**Comment by YaelDillies** [comment]
Oh yes, sorry, I overlooked that. `p` should be implicit. Can you change it?

**Comment by niklasmohrin** [comment]
Just so we are on the same page because I cannot make out a clear answer to the question in my comment: I now changed _my_ definition like you said, but left the existing definition `Walk.mem_finsetWalkLength_iff_length_eq` as before (with `Walk.` namespace and explicit `p`). This means that the definitions are inconsistent with each other, although they are the same statement just for the length-eq and the length-lt cases. If that is not what you meant, please let me know

**Comment by YaelDillies** [comment]
You should feel free to make `p` implicit in the existing one as well

**Comment by jcommelin** [comment]
Since there's still discussion going on, I've removed the `maintainer-merge` label for now.

**Comment by YaelDillies** [comment]
Okay, I think we're good?

maintainer merge

**Comment by semorrison** [comment]
bors merge

**Comment by kbuzzard** [comment]
I was just reading through this -- happy with the merge, but I personally am not clear about where in the library we're using `Fintype` and where we're using `Finite`. For example in finite group theory we seem to be moving to `Finite`. Why is `Fintype` the preferred option here?

**Comment by niklasmohrin** [comment]
The way I understand it is that with Fintype one can actually compute the elements (because this computation is provided with `elems`) whereas Finite only contains a proof. Since some downstream algorithm may want to enumerate all paths, Fintype seems to be the right choice here. I am lost on when to use Finite vs noncomputable Fintype though

**Pull request closed without merging**
